File: go/spore-core/sandbox.go
// Canonical SandboxProvider implementation — issue #6.
//
// WorkspaceScopedSandbox enforces a workspace root with allow/deny lists,
// extension filters, a read-only mode, and per-file size limits. It runs
// subprocesses directly via os/exec.CommandContext and offloads large
// outputs to {workspace_root}/.spore/offload/{call_id}.txt.

package sporecore

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// ============================================================================
// BuildError
// ============================================================================

// BuildErrorKind discriminates BuildError variants.
type BuildErrorKind string

const (
	BuildErrRootNotFound      BuildErrorKind = "root_not_found"
	BuildErrRootIO            BuildErrorKind = "root_io"
	BuildErrWriteRootNotFound BuildErrorKind = "write_root_not_found"
	BuildErrWriteRootIO       BuildErrorKind = "write_root_io"
)

// BuildError is the construction-time error type for WorkspaceScopedSandbox.
type BuildError struct {
	Kind BuildErrorKind
	Path string
	Err  error
}

func (e *BuildError) Error() string {
	switch e.Kind {
	case BuildErrRootNotFound:
		return fmt.Sprintf("workspace root does not exist: %s", e.Path)
	case BuildErrRootIO:
		return fmt.Sprintf("workspace root io error: %s: %s", e.Path, e.Err)
	case BuildErrWriteRootNotFound:
		return fmt.Sprintf("workspace write_root does not exist: %s", e.Path)
	case BuildErrWriteRootIO:
		return fmt.Sprintf("workspace write_root io error: %s: %s", e.Path, e.Err)
	default:
		return fmt.Sprintf("workspace build error: %s", e.Path)
	}
}

func (e *BuildError) Unwrap() error { return e.Err }

// ============================================================================
// WorkspaceScopedSandbox
// ============================================================================

// WorkspaceScopedSandbox is the canonical path-enforcing sandbox.
type WorkspaceScopedSandbox struct {
	config        WorkspaceConfig // .Root is the canonical (abs + symlink-resolved) root
	isolationMode IsolationMode
}

// NewWorkspaceScopedSandbox builds a sandbox with IsolationWorkspaceScoped.
func NewWorkspaceScopedSandbox(cfg WorkspaceConfig) (*WorkspaceScopedSandbox, error) {
	return NewWorkspaceScopedSandboxWithMode(cfg, IsolationWorkspaceScoped{})
}

// warnIfDangerousIsolation is wired up by the `dangerous` build to emit a
// construction-time warning when a sandbox is built with IsolationNone. It is
// nil in the default build, where IsolationNone does not exist.
var warnIfDangerousIsolation func(mode IsolationMode)

// NewWorkspaceScopedSandboxWithMode builds a sandbox with an explicit
// isolation mode. Under the `dangerous` build tag, IsolationNone emits a
// warning via log.Printf — it must never be enabled silently in production.
func NewWorkspaceScopedSandboxWithMode(cfg WorkspaceConfig, mode IsolationMode) (*WorkspaceScopedSandbox, error) {
	if cfg.Root == "" {
		return nil, &BuildError{Kind: BuildErrRootNotFound, Path: cfg.Root}
	}
	abs, err := filepath.Abs(cfg.Root)
	if err != nil {
		return nil, &BuildError{Kind: BuildErrRootIO, Path: cfg.Root, Err: err}
	}
	canonical, err := filepath.EvalSymlinks(abs)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, &BuildError{Kind: BuildErrRootNotFound, Path: cfg.Root}
		}
		return nil, &BuildError{Kind: BuildErrRootIO, Path: cfg.Root, Err: err}
	}
	info, err := os.Stat(canonical)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, &BuildError{Kind: BuildErrRootNotFound, Path: cfg.Root}
		}
		return nil, &BuildError{Kind: BuildErrRootIO, Path: cfg.Root, Err: err}
	}
	if !info.IsDir() {
		return nil, &BuildError{Kind: BuildErrRootIO, Path: cfg.Root, Err: fmt.Errorf("not a directory")}
	}
	cfg.Root = canonical

	// SC-13: canonicalize the optional write_root the same way so the write
	// boundary check compares canonical paths. It must exist (typically a
	// subdirectory of root). The path strings callers pass stay root-relative;
	// write_root only narrows the boundary, it is not a second join base.
	if cfg.WriteRoot != "" {
		wrAbs, err := filepath.Abs(cfg.WriteRoot)
		if err != nil {
			return nil, &BuildError{Kind: BuildErrWriteRootIO, Path: cfg.WriteRoot, Err: err}
		}
		wrCanonical, err := filepath.EvalSymlinks(wrAbs)
		if err != nil {
			if errors.Is(err, os.ErrNotExist) {
				return nil, &BuildError{Kind: BuildErrWriteRootNotFound, Path: cfg.WriteRoot}
			}
			return nil, &BuildError{Kind: BuildErrWriteRootIO, Path: cfg.WriteRoot, Err: err}
		}
		cfg.WriteRoot = wrCanonical
	}

	if warnIfDangerousIsolation != nil {
		warnIfDangerousIsolation(mode)
	}
	return &WorkspaceScopedSandbox{config: cfg, isolationMode: mode}, nil
}

// Config returns the (canonicalized) workspace config.
func (s *WorkspaceScopedSandbox) Config() WorkspaceConfig { return s.config }

// IsolationMode returns the active isolation mode.
func (s *WorkspaceScopedSandbox) IsolationMode() IsolationMode { return s.isolationMode }

// WorkspaceRoot returns the canonical workspace root.
func (s *WorkspaceScopedSandbox) WorkspaceRoot() string { return s.config.Root }

// Validate accepts every call — path resolution / policy enforcement happens
// in ResolvePath and ExecuteCommand.
func (s *WorkspaceScopedSandbox) Validate(context.Context, ToolCall) *SandboxViolation {
	return nil
}

// ============================================================================
// Path resolution — eight steps as specified.
// ============================================================================

// ResolvePath canonicalizes path against the workspace root and validates
// against allow/deny/extension/read-only policy for the given operation.
func (s *WorkspaceScopedSandbox) ResolvePath(_ context.Context, raw string, op Operation) (string, *SandboxViolation) {
	// 1. Join raw onto root, treating absolute paths as relative.
	rawPath := raw
	if filepath.IsAbs(rawPath) {
		// Strip a single leading separator so absolute callers don't
		// silently escape — `/foo` becomes `<root>/foo`.
		rawPath = strings.TrimPrefix(rawPath, string(filepath.Separator))
	}
	joined := filepath.Join(s.config.Root, rawPath)

	// 2. Canonicalize. The target file may not yet exist — for *any*
	// operation, including Read — so canonicalize the parent and re-join the
	// leaf. Resolution is operation-agnostic on purpose: existence is
	// orthogonal to the boundary check. A missing in-workspace path still
	// resolves (via its canonicalized parent) and passes the boundary check;
	// the actual read then naturally returns NotFound, surfaced as a
	// recoverable error by the read tool rather than a PathEscape. A missing
	// path that resolves *outside* the root is still a PathEscape.
	canonical, err := canonicalize(joined)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			parent := filepath.Dir(joined)
			leaf := filepath.Base(joined)
			cp, parentErr := canonicalize(parent)
			if parentErr != nil {
				return "", &SandboxViolation{Kind: SandboxPathEscape, Path: raw}
			}
			canonical = filepath.Join(cp, leaf)
		} else {
			return "", &SandboxViolation{Kind: SandboxPathEscape, Path: raw}
		}
	}

	// 3. Boundary check. Reads and execute must stay under the read Root;
	//    writes must stay under WriteRoot when set (SC-13: read-everywhere,
	//    write-scoped). The path string was already joined onto Root above —
	//    WriteRoot only narrows where a resolved write target may land, it is
	//    NOT a separate join base — so a write outside WriteRoot is a PathEscape
	//    even though it lives under Root.
	boundary := s.config.Root
	if op == OperationWrite && s.config.WriteRoot != "" {
		boundary = s.config.WriteRoot
	}
	if !pathHasPrefix(canonical, boundary) {
		return "", &SandboxViolation{Kind: SandboxPathEscape, Path: canonical}
	}

	// 4. Denylist (relative to root).
	for _, denied := range s.config.DeniedPaths {
		deniedAbs := filepath.Join(s.config.Root, denied)
		if pathHasPrefix(canonical, deniedAbs) {
			return "", &SandboxViolation{
				Kind:        SandboxPathDenied,
				Path:        canonical,
				MatchedRule: denied,
			}
		}
	}

	// 5. Allowlist (if non-empty).
	if len(s.config.AllowedPaths) > 0 {
		ok := false
		for _, allowed := range s.config.AllowedPaths {
			allowedAbs := filepath.Join(s.config.Root, allowed)
			if pathHasPrefix(canonical, allowedAbs) {
				ok = true
				break
			}
		}
		if !ok {
			return "", &SandboxViolation{
				Kind:        SandboxPathDenied,
				Path:        canonical,
				MatchedRule: "not in allowlist",
			}
		}
	}

	// 6. Denied extensions. Match either the path's extension or a dotfile
	// whose name (minus the leading dot) equals the denied extension.
	extFromExt := ""
	if e := filepath.Ext(canonical); e != "" {
		extFromExt = strings.ToLower(strings.TrimPrefix(e, "."))
	}
	base := filepath.Base(canonical)
	extFromDotfile := ""
	if strings.HasPrefix(base, ".") && !strings.Contains(base[1:], ".") {
		extFromDotfile = strings.ToLower(base[1:])
	}
	for _, deniedExt := range s.config.DeniedExtensions {
		trimmed := strings.ToLower(strings.TrimPrefix(deniedExt, "."))
		if trimmed == "" {
			continue
		}
		if extFromExt == trimmed || extFromDotfile == trimmed {
			return "", &SandboxViolation{
				Kind:      SandboxExtensionDenied,
				Path:      canonical,
				Extension: trimmed,
			}
		}
	}

	// 7. Read-only check.
	if s.config.ReadOnly && (op == OperationWrite || op == OperationExecute) {
		return "", &SandboxViolation{Kind: SandboxReadOnly, Path: canonical}
	}

	// 8. File-size cap (read only; writes are sized by content length upstream).
	if op == OperationRead && s.config.MaxFileSize > 0 {
		if info, err := os.Stat(canonical); err == nil && !info.IsDir() {
			if uint64(info.Size()) > s.config.MaxFileSize {
				return "", &SandboxViolation{
					Kind:  SandboxFileSizeExceeded,
					Path:  canonical,
					Size:  uint64(info.Size()),
					Limit: s.config.MaxFileSize,
				}
			}
		}
	}

	return canonical, nil
}

// canonicalize resolves a path to an absolute, symlink-free form. Returns
// the underlying os error (so callers can test for os.ErrNotExist) without
// wrapping.
func canonicalize(p string) (string, error) {
	abs, err := filepath.Abs(p)
	if err != nil {
		return "", err
	}
	return filepath.EvalSymlinks(abs)
}

// pathHasPrefix reports whether path is rooted at prefix. It is component-
// aware so /foo/bar is not considered a prefix of /foo/barbaz.
func pathHasPrefix(path, prefix string) bool {
	path = filepath.Clean(path)
	prefix = filepath.Clean(prefix)
	if path == prefix {
		return true
	}
	sep := string(filepath.Separator)
	if !strings.HasSuffix(prefix, sep) {
		prefix += sep
	}
	return strings.HasPrefix(path+sep, prefix)
}

// ============================================================================
// ExecuteCommand
// ============================================================================

// ExecuteCommand runs a subprocess inside the configured isolation mode.
// Bubblewrap/Docker modes return DisallowedCommand until those backends are
// wired up.
func (s *WorkspaceScopedSandbox) ExecuteCommand(
	ctx context.Context,
	command string,
	args []string,
	workingDir string,
	timeout time.Duration,
) (CommandOutput, *SandboxViolation) {
	switch s.isolationMode.(type) {
	case IsolationBubblewrap:
		return CommandOutput{}, &SandboxViolation{
			Kind:    SandboxDisallowedCommand,
			Command: fmt.Sprintf("bubblewrap isolation not implemented: %s", command),
		}
	case IsolationDocker:
		return CommandOutput{}, &SandboxViolation{
			Kind:    SandboxDisallowedCommand,
			Command: fmt.Sprintf("docker isolation not implemented: %s", command),
		}
	default:
		// IsolationWorkspaceScoped (and, under the `dangerous` build tag,
		// IsolationNone) proceed directly.
	}

	// SC-12: apply exec-hardening knobs when configured. The per-call timeout
	// always wins; DefaultTimeout is the floor for callers that pass none (0).
	effectiveTimeout := timeout
	if effectiveTimeout == 0 && s.config.ExecConfig != nil {
		effectiveTimeout = s.config.ExecConfig.DefaultTimeout
	}

	runCtx := ctx
	var cancel context.CancelFunc
	if effectiveTimeout > 0 {
		runCtx, cancel = context.WithTimeout(ctx, effectiveTimeout)
		defer cancel()
	}
	// KillOnDrop is satisfied by construction: exec.CommandContext kills the
	// child when runCtx is cancelled (the timeout firing, or an outer cancel),
	// which is Go's native kill-on-drop. No extra wiring is needed.
	cmd := exec.CommandContext(runCtx, command, args...)
	if workingDir != "" {
		cmd.Dir = workingDir
	} else {
		cmd.Dir = s.config.Root
	}

	// SC-12: CloseStdin and NonInteractiveEnv are gated on ExecConfig so the
	// legacy (nil) path is byte-identical (inherited stdin, inherited env).
	if ec := s.config.ExecConfig; ec != nil {
		if ec.CloseStdin {
			// An empty reader yields immediate EOF, so input-blocked commands
			// fail fast instead of hanging on the inherited terminal.
			cmd.Stdin = bytes.NewReader(nil)
		}
		if len(ec.NonInteractiveEnv) > 0 {
			// Force vars onto the inherited environment in sorted key order
			// (deterministic), matching the Rust reference's BTreeMap.
			env := os.Environ()
			keys := make([]string, 0, len(ec.NonInteractiveEnv))
			for k := range ec.NonInteractiveEnv {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			for _, k := range keys {
				env = append(env, k+"="+ec.NonInteractiveEnv[k])
			}
			cmd.Env = env
		}
	}

	stdout, err := cmd.Output()
	stderr := ""
	if ee, ok := err.(*exec.ExitError); ok {
		stderr = string(ee.Stderr)
	}
	exitCode := 0
	if err != nil {
		if runCtx.Err() == context.DeadlineExceeded {
			return CommandOutput{
				Stdout:   "",
				Stderr:   fmt.Sprintf("command timed out after %s", effectiveTimeout),
				ExitCode: -1,
				TimedOut: true,
			}, nil
		}
		if ee, ok := err.(*exec.ExitError); ok {
			exitCode = ee.ExitCode()
		} else {
			// SC-15: a failed spawn is a typed violation, not a fake
			// CommandOutput{ExitCode: -1}. Callers already handle the
			// *SandboxViolation arm. Timeout (handled above) keeps its
			// CommandOutput{ExitCode: -1, TimedOut: true} — a real run that
			// exceeded the clock, not a spawn failure.
			return CommandOutput{}, &SandboxViolation{
				Kind:    SandboxExecSpawnFailed,
				Command: command,
				Message: err.Error(),
			}
		}
	}
	return CommandOutput{
		Stdout:   string(stdout),
		Stderr:   stderr,
		ExitCode: exitCode,
	}, nil
}

// ============================================================================
// HandleLargeOutput
// ============================================================================

// HandleLargeOutput head+tail-truncates content and offloads the original
// to {workspace_root}/.spore/offload/{call_id}.txt.
func (s *WorkspaceScopedSandbox) HandleLargeOutput(
	_ context.Context,
	content string,
	callID string,
	headTokens uint32,
	tailTokens uint32,
) TruncatedOutput {
	headChars := int(headTokens) * 4
	tailChars := int(tailTokens) * 4
	runes := []rune(content)
	originalSize := uint64(len(content))
	if len(runes) <= headChars+tailChars {
		return TruncatedOutput{Content: content, Truncated: false, OriginalSize: originalSize}
	}
	head := string(runes[:headChars])
	tail := string(runes[len(runes)-tailChars:])
	snippet := fmt.Sprintf("%s\n...[truncated]...\n%s", head, tail)

	// Offload the full content.
	var fullRef *FileRef
	offloadDir := filepath.Join(s.config.Root, ".spore", "offload")
	if err := os.MkdirAll(offloadDir, 0o755); err == nil {
		safeID := sanitizeCallID(callID)
		offloadPath := filepath.Join(offloadDir, safeID+".txt")
		if err := os.WriteFile(offloadPath, []byte(content), 0o644); err == nil {
			fullRef = &FileRef{Path: offloadPath, ByteLen: originalSize}
		}
	}

	return TruncatedOutput{
		Content:      snippet,
		Truncated:    true,
		FullRef:      fullRef,
		OriginalSize: originalSize,
	}
}

func sanitizeCallID(id string) string {
	var b strings.Builder
	b.Grow(len(id))
	for _, c := range id {
		switch {
		case c >= '0' && c <= '9', c >= 'a' && c <= 'z', c >= 'A' && c <= 'Z', c == '-', c == '_':
			b.WriteRune(c)
		default:
			b.WriteRune('_')
		}
	}
	return b.String()
}

// Compile-time check.
var _ SandboxProvider = (*WorkspaceScopedSandbox)(nil)

File: python/packages/spore_core/src/spore_core/sandbox.py
"""Canonical :class:`SandboxProvider` implementation — issue #6.

:class:`WorkspaceScopedSandbox` is the default in-tree sandbox. It enforces a
workspace root with allow/deny lists, extension filters, a read-only mode,
and per-file size limits. It runs subprocesses directly via
:func:`asyncio.create_subprocess_exec` and offloads large outputs to
``{workspace_root}/.spore/offload/{call_id}.txt``.

Mirrors ``rust/crates/spore-core/src/sandbox.rs`` byte-for-byte where the wire
format crosses the language boundary.
"""

from __future__ import annotations

import asyncio
import logging
import os
import warnings
from pathlib import Path

from .errors import SporeError
from .harness import (
    CommandOutput,
    FileRef,
    IsolationMode,
    IsolationModeBubblewrap,
    IsolationModeDocker,
    IsolationModeNone,
    IsolationModeWorkspaceScoped,
    Operation,
    SandboxDisallowedCommand,
    SandboxExecSpawnFailed,
    SandboxExtensionDenied,
    SandboxFileSizeExceeded,
    SandboxPathDenied,
    SandboxPathEscape,
    SandboxReadOnlyViolation,
    SandboxViolation,
    TruncatedOutput,
    WorkspaceConfig,
)
from .model import ToolCall

_logger = logging.getLogger(__name__)


class SandboxBuildError(SporeError):
    """Construction-time error for :class:`WorkspaceScopedSandbox`."""


class SandboxViolationException(SporeError):
    """Wraps a :class:`SandboxViolation` when raised as an exception.

    The canonical Python contract is to return violations as values; this
    type exists so internal helpers can short-circuit through ``try/except``
    when convenient.
    """

    def __init__(self, violation: SandboxViolation) -> None:
        super().__init__(f"sandbox violation: {violation.kind}")
        self.violation = violation


# ============================================================================
# WorkspaceScopedSandbox
# ============================================================================


class WorkspaceScopedSandbox:
    """Path-enforcing sandbox.

    Canonicalizes every resolved path against the workspace root and applies
    allow/deny + extension + read-only policies. Implements the
    :class:`SandboxProvider` protocol structurally — does not inherit.
    """

    def __init__(
        self,
        config: WorkspaceConfig,
        mode: IsolationMode | None = None,
    ) -> None:
        # Validate root exists, then canonicalize for stable comparisons.
        try:
            canonical = config.root.resolve(strict=True)
        except FileNotFoundError as e:
            raise SandboxBuildError(f"workspace root does not exist: {config.root}") from e
        except OSError as e:
            raise SandboxBuildError(f"workspace root io error: {config.root}: {e}") from e

        # SC-13: canonicalize the optional write_root the same way so the
        # boundary check compares canonical paths. It must already exist.
        canonical_write_root: Path | None = None
        if config.write_root is not None:
            try:
                canonical_write_root = config.write_root.resolve(strict=True)
            except FileNotFoundError as e:
                raise SandboxBuildError(
                    f"workspace write_root does not exist: {config.write_root}"
                ) from e
            except OSError as e:
                raise SandboxBuildError(
                    f"workspace write_root io error: {config.write_root}: {e}"
                ) from e

        self._config = config.model_copy(
            update={
                "root": canonical,
                "allowed_paths": [
                    self._normalize_under(canonical, p) for p in config.allowed_paths
                ],
                "denied_paths": [self._normalize_under(canonical, p) for p in config.denied_paths],
                "write_root": canonical_write_root,
            }
        )
        self._mode: IsolationMode = mode or IsolationModeWorkspaceScoped()

        if isinstance(self._mode, IsolationModeNone):
            msg = (
                "WorkspaceScopedSandbox constructed with IsolationMode=None — "
                "trusted-dev use only; do not enable silently in production"
            )
            warnings.warn(msg, stacklevel=2)
            _logger.warning(msg)

    @staticmethod
    def _normalize_under(root: Path, p: Path) -> Path:
        """Anchor a configured allow/deny path under the canonical root."""

        candidate = p if p.is_absolute() else root / p
        # Use resolve(strict=False) so we tolerate not-yet-existing entries.
        return candidate.resolve(strict=False)

    @property
    def config(self) -> WorkspaceConfig:
        return self._config

    # ---- protocol surface -------------------------------------------------

    def isolation_mode(self) -> IsolationMode:
        return self._mode

    def workspace_root(self) -> Path:
        return self._config.root

    async def validate(self, call: ToolCall) -> SandboxViolation | None:
        # Per-tool sandbox validation is performed at the call sites via
        # ``resolve_path``; this hook stays as a no-op pass-through. Tools
        # that don't go through ``resolve_path`` (e.g. ``bash_command``) are
        # gated by ``execute_command``.
        _ = call
        return None

    async def resolve_path(self, path: str, operation: Operation = "read") -> Path:
        result = self._resolve(path, operation)
        if isinstance(result, Path):
            return result
        raise SandboxViolationException(result)

    # ---- core resolution --------------------------------------------------

    def _resolve(self, raw: str, operation: Operation) -> Path | SandboxViolation:
        """Path-resolution core. Returns either a canonical :class:`Path`
        or a :class:`SandboxViolation` value.

        Follows the 8-step algorithm in the spec.
        """

        root = self._config.root

        # 1. Join root + raw, treating absolute paths as relative-to-root.
        raw_path = Path(raw)
        if raw_path.is_absolute():
            # Strip the leading separator(s).
            stripped = Path(*raw_path.parts[1:]) if len(raw_path.parts) > 1 else Path()
            joined = root / stripped
        else:
            joined = root / raw_path

        # 2. Canonicalize. The target file may not yet exist — for *any*
        #    operation, including Read — so canonicalize the parent and
        #    re-join the filename. Resolution is operation-agnostic on
        #    purpose: existence is orthogonal to the boundary check. A missing
        #    in-workspace path still resolves (via its canonicalized parent)
        #    and passes the boundary check; the actual read then naturally
        #    returns NotFound, surfaced as a recoverable error by the read
        #    tool rather than a PathEscape. A missing path that resolves
        #    *outside* the root is still a PathEscape (the boundary check in
        #    step 3 rejects it).
        canonical: Path
        try:
            canonical = joined.resolve(strict=True)
        except FileNotFoundError:
            parent = joined.parent
            try:
                parent_canon = parent.resolve(strict=True)
            except (FileNotFoundError, OSError):
                return SandboxPathEscape(path=raw)
            file_name = joined.name
            if not file_name:
                return SandboxPathEscape(path=raw)
            canonical = parent_canon / file_name
        except OSError:
            return SandboxPathEscape(path=raw)

        # 3. Boundary check. Reads and execute must stay under the read `root`;
        #    writes must stay under `write_root` when set (SC-13:
        #    read-everywhere, write-scoped). The path string was already joined
        #    onto `root` above — `write_root` only narrows where the resolved
        #    target is allowed to land, it is NOT a separate join base — so a
        #    write outside `write_root` is a PathEscape even though it lives
        #    under `root`.
        if operation == "write" and self._config.write_root is not None:
            boundary = self._config.write_root
        else:
            boundary = root
        try:
            canonical.relative_to(boundary)
        except ValueError:
            return SandboxPathEscape(path=str(canonical))

        # 4. Denylist.
        for denied in self._config.denied_paths:
            try:
                canonical.relative_to(denied)
            except ValueError:
                continue
            return SandboxPathDenied(path=str(canonical), matched_rule=str(denied))

        # 5. Allowlist (if non-empty).
        if self._config.allowed_paths:
            in_allowlist = False
            for allowed in self._config.allowed_paths:
                try:
                    canonical.relative_to(allowed)
                except ValueError:
                    continue
                in_allowlist = True
                break
            if not in_allowlist:
                return SandboxPathDenied(path=str(canonical), matched_rule="not in allowlist")

        # 6. Denied extensions. Match the path's real extension or a dotfile
        #    name (e.g. ``.env`` → "env").
        ext_real: str | None = None
        if canonical.suffix:
            ext_real = canonical.suffix.lstrip(".").lower()
        ext_dotfile: str | None = None
        name = canonical.name
        if name.startswith(".") and "." not in name[1:]:
            ext_dotfile = name[1:].lower()
        for denied_ext in self._config.denied_extensions:
            trimmed = denied_ext.lstrip(".").lower()
            if ext_real == trimmed or ext_dotfile == trimmed:
                return SandboxExtensionDenied(path=str(canonical), extension=trimmed)

        # 7. Read-only.
        if self._config.read_only and operation in ("write", "execute"):
            return SandboxReadOnlyViolation(path=str(canonical))

        # 8. File-size cap (read ops only).
        if operation == "read" and self._config.max_file_size > 0 and canonical.is_file():
            try:
                size = canonical.stat().st_size
            except OSError:
                size = None
            if size is not None and size > self._config.max_file_size:
                return SandboxFileSizeExceeded(
                    path=str(canonical), size=size, limit=self._config.max_file_size
                )

        return canonical

    # ---- command execution ------------------------------------------------

    async def execute_command(
        self,
        command: str,
        args: list[str],
        working_dir: Path | None = None,
        timeout: float | None = None,
    ) -> CommandOutput:
        if isinstance(self._mode, IsolationModeBubblewrap):
            raise SandboxViolationException(
                SandboxDisallowedCommand(command=f"bubblewrap isolation not implemented: {command}")
            )
        if isinstance(self._mode, IsolationModeDocker):
            raise SandboxViolationException(
                SandboxDisallowedCommand(command=f"docker isolation not implemented: {command}")
            )

        cwd = str(working_dir) if working_dir is not None else str(self._config.root)

        # SC-12: apply exec-hardening knobs when configured. The per-call
        # `timeout` always wins; `default_timeout` is the floor for callers that
        # pass `None`. `None` exec_config keeps the legacy spawn byte-identical.
        ec = self._config.exec_config
        spawn_kwargs: dict[str, object] = {}
        effective_timeout = timeout
        kill_on_drop = False
        if ec is not None:
            if ec.close_stdin:
                spawn_kwargs["stdin"] = asyncio.subprocess.DEVNULL
            if ec.non_interactive_env:
                # `env=` REPLACES the environment, so merge the forced vars on top
                # of the inherited one. Applied in SORTED key order (deterministic,
                # matching Rust's BTreeMap).
                merged_env = dict(os.environ)
                for k in sorted(ec.non_interactive_env):
                    merged_env[k] = ec.non_interactive_env[k]
                spawn_kwargs["env"] = merged_env
            kill_on_drop = ec.kill_on_drop
            if timeout is None:
                effective_timeout = ec.default_timeout

        try:
            proc = await asyncio.create_subprocess_exec(
                command,
                *args,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                cwd=cwd,
                **spawn_kwargs,  # type: ignore[arg-type]
            )
        except (FileNotFoundError, OSError) as e:
            # SC-15: a failed spawn is a typed violation, not a fake
            # exit_code: -1 success. Callers already handle the exception.
            raise SandboxViolationException(
                SandboxExecSpawnFailed(command=command, message=str(e))
            ) from e

        try:
            if effective_timeout is not None:
                stdout_b, stderr_b = await asyncio.wait_for(
                    proc.communicate(), timeout=effective_timeout
                )
            else:
                stdout_b, stderr_b = await proc.communicate()
        except asyncio.TimeoutError:
            try:
                proc.kill()
            except ProcessLookupError:
                pass
            try:
                await proc.wait()
            except (ProcessLookupError, asyncio.CancelledError):
                pass
            secs = int(effective_timeout) if effective_timeout is not None else 0
            return CommandOutput(
                stdout="",
                stderr=f"command timed out after {secs}s",
                exit_code=-1,
                timed_out=True,
                truncated=False,
            )
        except asyncio.CancelledError:
            # SC-12 kill_on_drop: the execute_command future was cancelled. Reap
            # the child instead of orphaning it, then re-raise so cancellation
            # still propagates. Without the knob the child is left running
            # (legacy leak-on-drop behavior).
            if kill_on_drop:
                try:
                    proc.kill()
                except ProcessLookupError:
                    pass
            raise

        return CommandOutput(
            stdout=stdout_b.decode("utf-8", errors="replace"),
            stderr=stderr_b.decode("utf-8", errors="replace"),
            exit_code=proc.returncode if proc.returncode is not None else -1,
            timed_out=False,
            truncated=False,
        )

    # ---- large output offload --------------------------------------------

    async def handle_large_output(
        self,
        content: str,
        call_id: str,
        head_tokens: int,
        tail_tokens: int,
    ) -> TruncatedOutput:
        head_chars = max(0, head_tokens) * 4
        tail_chars = max(0, tail_tokens) * 4
        total_chars = len(content)
        original_size = len(content.encode("utf-8"))

        if total_chars <= head_chars + tail_chars:
            return TruncatedOutput(
                content=content,
                truncated=False,
                full_ref=None,
                original_size=original_size,
            )

        head = content[:head_chars]
        tail = content[total_chars - tail_chars :]
        snippet = f"{head}\n...[truncated]...\n{tail}"

        offload_dir = self._config.root / ".spore" / "offload"
        full_ref: FileRef | None = None
        try:
            offload_dir.mkdir(parents=True, exist_ok=True)
            safe_id = _sanitize_call_id(call_id)
            offload_path = offload_dir / f"{safe_id}.txt"
            offload_path.write_text(content, encoding="utf-8")
            full_ref = FileRef(path=str(offload_path), byte_len=original_size)
        except OSError:
            full_ref = None

        return TruncatedOutput(
            content=snippet,
            truncated=True,
            full_ref=full_ref,
            original_size=original_size,
        )


def _sanitize_call_id(call_id: str) -> str:
    return "".join(c if (c.isalnum() or c in ("-", "_")) else "_" for c in call_id)


__all__ = [
    "SandboxBuildError",
    "SandboxViolationException",
    "WorkspaceScopedSandbox",
]
